datatype/list.go

```diff
@@ -150,4 +176,40 @@ func writeCollectionSize(version primitive.ProtocolVersion, n int, buf *bytes.Bu
 
 		return primitive.WriteShort(uint16(n), buf)
 	}
+}
+
+func readCollectionSize(version primitive.ProtocolVersion, encoded []byte) (int, int, error) {
+	if version.Uses4BytesCollectionLength() {
+		sizeInt32, err := primitive.ReadInt(bytes.NewBuffer(encoded))
+		return int(sizeInt32), 4, err
+	} else {
+		sizeInt16, err := primitive.ReadShort(bytes.NewBuffer(encoded))
+		return int(sizeInt16), 2, err
+	}
+}
+
+func decodeChildElement(
+	elementCodec Codec,
+	elementType reflect.Type,
+	encoded []byte,
+	version primitive.ProtocolVersion) (output *reflect.Value, read int, err error) {
+	m, read, err := readCollectionSize(version, encoded)
```

Comment:
Why are we reading the collection size again here? We should instead read the element size – but that's already done inside `elementCodec.Decode`. Actually I'm not sure I follow why we have this method, I think a simple call to `elementCodec.Decode` should suffice.

Reply:
I think the problem might stem from the usage of slices in the API:

```
	Encode(value interface{}, version primitive.ProtocolVersion) (encoded []byte, err error)
	Decode(encoded []byte, version primitive.ProtocolVersion) (value interface{}, err error)
```

I used slices because indeed that's what we have when decoding a `Rows` struct, for instance.

But I'm realizing that with slices it's hard to keep track of what has been decoded already, and where is the cursor once the flow returns from the `Encode` or `Decode` methods.

Taking a step back, wouldn't it be better if we replaced the slices with `io.Writer` and `io.Reader` as we did for the rest of the library?

Reply:
>>>Why are we reading the collection size again here? We should instead read the element size – but that's already done inside elementCodec.Decode. Actually I'm not sure I follow why we have this method, I think a simple call to elementCodec.Decode should suffice.

I don't think this is true? None of the current codecs read the element size. I think the element size is not parsed into the `RowsResult` object.

> Taking a step back, wouldn't it be better if we replaced the slices with `io.Writer` and `io.Reader` as we did for the rest of the library?

I started doing that but then I went back but I can't remember exactly why I did that... I could give it a try again but I don't think there is a very clear benefit to using that. The current implementation is working nicely from the tests that I've ran so far.

Reply:
>This is more what I had in mind for the general logic of a collection codec:

Codecs don't read the sizes though, at least they don't atm.